src/java/org/apache/cassandra/index/sai/disk/format/IndexDescriptor.java

```diff
@@ -0,0 +1,530 @@
+    }
+
+    public Set<Component> getLivePerIndexComponents(IndexContext indexContext)
+    {
+        registerPerIndexComponents(indexContext);
+        return perIndexComponents.containsKey(indexContext.getIndexName())
+               ? perIndexComponents.get(indexContext.getIndexName())
+                                   .stream()
+                                   .map(c -> new Component(Component.Type.CUSTOM, componentName(c, indexContext)))
+                                                                         .collect(Collectors.toSet())
+                                                     : Collections.emptySet();
+    }
+
+    public PrimaryKeyMap.Factory newPrimaryKeyMapFactory(SSTableReader sstable)
+    {
+        return version.onDiskFormat().newPrimaryKeyMapFactory(this, sstable);
+    }
+
+    public SearchableIndex newSearchableIndex(SSTableContext sstableContext, IndexContext indexContext)
+    {
+        return version.onDiskFormat().newSearchableIndex(sstableContext, indexContext);
+    }
+
+    public PerSSTableWriter newPerSSTableWriter() throws IOException
+    {
+        return version.onDiskFormat().newPerSSTableWriter(this);
+    }
+
+    public PerIndexWriter newPerIndexWriter(StorageAttachedIndex index,
+                                            LifecycleNewTracker tracker,
+                                            RowMapping rowMapping)
+    {
+        return version.onDiskFormat().newPerIndexWriter(index, this, tracker, rowMapping);
+    }
+
+    public boolean isPerSSTableBuildComplete()
+    {
+        return version.onDiskFormat().isPerSSTableBuildComplete(this);
+    }
+
+    public boolean isPerIndexBuildComplete(IndexContext indexContext)
+    {
+        return version.onDiskFormat().isPerIndexBuildComplete(this, indexContext);
+    }
+
+    public boolean isIndexEmpty(IndexContext indexContext)
+    {
+        return isPerIndexBuildComplete(indexContext) && numberOfComponents(indexContext) == 1;
+    }
+
+    public long sizeOnDiskOfPerIndexComponents(IndexContext indexContext)
+    {
+        registerPerIndexComponents(indexContext);
+        if (perIndexComponents.containsKey(indexContext.getIndexName()))
+            return perIndexComponents.get(indexContext.getIndexName())
+                                     .stream()
+                                     .map(c -> Pair.create(c, indexContext.getIndexName()))
+                                     .map(onDiskPerIndexFileMap::get)
+                                     .filter(java.util.Objects::nonNull)
+                                     .filter(File::exists)
+                                     .mapToLong(File::length)
+                                     .sum();
+        return 0;
+    }
+
+    @VisibleForTesting
+    public long sizeOnDiskOfPerIndexComponent(IndexComponent indexComponent, IndexContext indexContext)
+    {
+        if (perIndexComponents.containsKey(indexContext.getIndexName()))
```

Comment:
Only used for testing, so no registration?

Reply:
There is no registration now but yes, it should have been calling register.